### PR TITLE
Weights & Biases

### PR DESCRIPTION
**Pull request by ramanshsharma2806**
We have used the [Weights & Biases](http://wandb.ai/) tool to facilitate several aspects of the policy incentive classification modeling process

All the data for our project can be found here: [https://wandb.ai/ramanshsharma/WRI](https://wandb.ai/ramanshsharma/WRI)

Some of the things that Weights & Biases has been used for is
1. Keeping track of every experiment run on fine-tuning SBERT for classification.
2. Keeping track of hyperparameters used in each run.
3. Analyzing automatically generated training/validation accuracy plots, F1 score plots for macro and weighted based on the information we log from the code.
4. Run automatic hyperparameter tuning using W&B sweeps. For example this one [https://wandb.ai/ramanshsharma/WRI/sweeps/fyc8q0fx](https://wandb.ai/ramanshsharma/WRI/sweeps/fyc8q0fx). We can visually see the combination of hyperparameters that lead to the best results in the model's F1 performance on the validation set in the Parallel coordinates plot. 
5. Save the models on Weights & Biases (**keeping in mind the 200 GB limit on stored information on the project**) and loading them in the notebook for inference.

Moving forward with more tasks this will be incredibly helpful in organizing large scale experiments with more data and more hyperparameters to tune.

### CODE REVIEW

File: tasks/fine_tuning_sbert/src/custom_evaluator.py [hunk 1]
```diff
@@ -110,8 +108,10 @@ def __init__(self, dataloader: DataLoader, name: str = "", label_names: list = N
             name = "_" + name
 
         self.csv_file = "accuracy_f1_evaluation" + name + "_results.csv"
-        self.csv_headers = [param_name for param_name in self.model_hyper_params]
-        self.csv_headers.extend(["epoch", "steps", "accuracy", "macro_f1", "weighted_f1"])
+        self.csv_headers = [
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/custom_evaluator.py:111]: We're not storing a csv anymore right? then let's get rid of these

File: tasks/fine_tuning_sbert/src/custom_evaluator.py [hunk 2]
```diff
@@ -74,13 +74,11 @@ def plot_confusion_matrix(cm, label_names, title='Confusion matrix',
...
-
-    plt.show()
+    # wandb code
+    fig = plt.gcf()
+    fig.set_size_inches(15, 10)  # enlarging CM
+    wandb.log({"confusion matrix": wandb.Image(fig)})
+    plt.close()  # this should prevent output of plt.imshow above
 
 
 class CustomLabelAccuracyEvaluator(SentenceEvaluator):
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/custom_evaluator.py:84]: Here, we're also not using self.model_hyper_params, so let's delete it.

File: tasks/fine_tuning_sbert/src/loops.py [hunk 3]
```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
...
+    warmup_steps = math.ceil(
+        len(train_dataset) * config.max_num_epochs / train_batch_size * 0.1)  # 10% of train data for warm-up
+
+    set_seeds(config.seeds)
+
+    # Train the model
+    start = time.time()
+    dev_evaluator = CustomLabelAccuracyEvaluator(dataloader=dev_dataloader, softmax_model=classifier,
+                                                 name='lae-dev', label_names=label_names,
+                                                 model_hyper_params={'model_name': config.model_name, 'dev_perc': config.dev_perc, 'seed': config.seeds})
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/loops.py:122]: Also here, no need for model_hyper_params :)

File: tasks/fine_tuning_sbert/src/loops.py [hunk 4]
```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
...
+
+    model.fit(train_objectives=[(train_dataloader, classifier)],
+              evaluator=dev_evaluator,
+              epochs=config.max_num_epochs,
+              evaluation_steps=1000,
+              warmup_steps=warmup_steps,
+              output_path=config.output_path,
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/loops.py:129]: Why is the output path part of the config? Can we leave it out so we don't optimize for it :D?

File: tasks/fine_tuning_sbert/src/loops.py [hunk 5]
```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
...
+              model_deets=model_deets,
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/loops.py:130]: Let's get rid of model deets here, as well as in `sentence_transformer.py` and in `custom_evaluator.py`, because that was ultimately sent to output the confusion matrix, which now gets stored in wandb

File: tasks/fine_tuning_sbert/src/loops.py [hunk 6]
```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
...
+    end = time.time()
+    hours, rem = divmod(end - start, 3600)
+    minutes, seconds = divmod(rem, 60)
+    print("Time taken for fine-tuning:",
+          "{:0>2}:{:0>2}:{:05.2f}".format(int(hours), int(minutes), seconds))
+
+    return run_name
+
+
+def make_dataset_public(train_sents_, train_labels_, label_names_):
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/loops.py:228]: I'm not gonna lie, I get scared of global variables but in this case, well done - it seems like a good solution for now :)

File: tasks/fine_tuning_sbert/src/loops.py [hunk 7]
```diff
@@ -212,13 +315,15 @@ def evaluate_using_sklearn(clf, model, train_sents, train_labels, test_sents, te
...
-
-    print("Visualizing...")
-    out_path = f"{output_path}/{model_deets}" if output_path and model_deets else None
-    visualize_embeddings_2D(np.vstack(test_embs), test_labels, tsne_perplexity=50,
-                            output_path=out_path)
-    evaluator.plot_confusion_matrix(color_map='Blues', output_path=out_path)
+    evaluator = ModelEvaluator(
+        label_names, y_true=numeric_test_labels, y_pred=numeric_preds)
+
+    if not testing:
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/loops.py:321]: Should we also store the CM, embedding visualization and classification report in wandb? or is that unnecessary? your call!

File: tasks/fine_tuning_sbert/src/sentence_transformer.py [hunk 8]
```diff
@@ -29,7 +30,8 @@ def fit(self,
             scheduler: str = 'WarmupLinear',
             warmup_steps: int = 10000,
             optimizer_class: Type[Optimizer] = transformers.AdamW,
-            optimizer_params: Dict[str, object] = {'lr': 2e-5, 'correct_bias': True},
+            optimizer_params: Dict[str, object] = {
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/sentence_transformer.py:33]: We don't need to do this now if it is too much work, but I'd suggest we find a way of including the `lr` parameter in the config file for the sweep... would that work? that way it sets a standard for whenever someone wants to optimize a hyperparameter in this function.

File: tasks/fine_tuning_sbert/src/sentence_transformer.py [hunk 9]
```diff
@@ -199,42 +212,37 @@ def fit(self,
...
-                self.plot_train_val_accuracy(model_deets, output_path, training_acc_list)
                 return
 
             # removing the unnecessary first element in ACC_LIST that needed to be there for epoch 1
             if epoch == 0:
                 del self.acc_list[0]
 
-        if evaluator is None and output_path is not None:  # No evaluator, but output path: save final model version
+        # No evaluator, but output path: save final model version
+        if evaluator is None and output_path is not None:
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/sentence_transformer.py:222]: We still do this first before saving into wandb right? just making sure we're not double saving the model

File: tasks/fine_tuning_sbert/src/sentence_transformer.py [hunk 10]
```diff
@@ -199,42 +212,37 @@ def fit(self,
...
+        wandb.log(
+            {"Macro F1 validation": score_dict['macro_f1'], "epoch": epoch})
+        wandb.log(
+            {"Weighted F1 validation": score_dict['weighted_f1'], "epoch": epoch})
+
         prev_score = self.acc_list[-2]
         moving_average = mean(self.acc_list[-self.patience - 1: -1])
 
-        print(f"{'=' * 60}\nCurrent Score is: {score}\nCurrent ACC_LIST is: {self.acc_list}")
+        print(
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/src/sentence_transformer.py:244]: Is there a particular reason why you didn't choose to log this with wandb? Just curious :)

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb [hunk 11]
```diff
@@ -1,3160 +1,2839 @@
...
       },
-      "text/plain": [
-       "HBox(children=(FloatProgress(value=0.0, description='Iteration', max=41.0, style=ProgressStyle(description_wid…"
+      "source": [
+        "# Install necessary libraries\n",
+        "! pip install --quiet \\\n",
+        "  scprep\\\n",
+        "  spacy==3.0.5 \\\n",
+        "  wandb \\\n",
+        "  onnx \\\n",
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1788]: Let's remove unused packages from here (onnx and onnxruntime), I'm not sure if we ever end up using spacy but if not, let's get rid of it.

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb [hunk 12]
```diff
@@ -1,3160 +1,2839 @@
...
+      "cell_type": "code",
+      "metadata": {
+        "id": "jm-fbjB9Rkzx"
+      },
+      "source": [
+        "\"\"\"\n",
+        "MAKE SURE THAT THIS IS ALWAYS YOUR INDIVIDUAL NUMBER (FOR REASONS TO NOT OVERWRITE DATA)\n",
+        "\"\"\"\n",
+        "\n",
+        "GLOBAL_EXPERIMENT_NUMBER = 16"
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1965]: If this is not going to matter in the future, let's join this cell with the next one and make it a random number. We'll delete the folder at the end anyway right?

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb [hunk 13]
```diff
@@ -1,3160 +1,2839 @@
...
-      "Epoch: 2\n",
-      "Best score: 0.9444444444444444\n",
-      "============================================================\n"
-     ]
+      "cell_type": "code",
+      "metadata": {
+        "id": "v1MTStwySrYL"
+      },
+      "source": [
+        "# PASTE YOUR WEIGHTS & BIASES KEY HERE\n",
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2086]: Let's add a comment saying that we should not forget to delete them, as I did :D

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb [hunk 14]
```diff
@@ -1,3160 +1,2839 @@
...
-       "version_major": 2,
-       "version_minor": 0
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "VbnqVcEzuWq5"
       },
-      "text/plain": [
-       "HBox(children=(FloatProgress(value=0.0, description='Iteration', max=41.0, style=ProgressStyle(description_wid…"
+      "source": [
+        "## Grid Search Fine Tuning - W&B sweeps"
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2104]: In this section and the next one (run training function for one run), I'd add some explanation/instructions on how to use a "sweep" and which cells to run, that way there's a clear distinction and everyone knows which cells to run when, and how to look at the results.

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb [hunk 15]
```diff
@@ -1,3160 +1,2839 @@
...
-       "version_major": 2,
-       "version_minor": 0
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "IT-nPon7lShD"
       },
-      "text/plain": [
-       "HBox(children=(FloatProgress(value=0.0, description='Iteration', max=41.0, style=ProgressStyle(description_wid…"
+      "source": [
+        "## Running training function for only one run"
```
Comment (thefirebanks) [inline tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2190]: If this is running it once, then it's not grid search anymore right? Could we rename the function please?

**Review by thefirebanks — CHANGES_REQUESTED** [review]
Amazingly done Ram! I know I wrote a lot of comments, but they are mainly just feedback to clean up the code as opposed to functionality changes. Apart from that, thank you for successfully bringing wandb to the project and I can't wait to play with it even more :D

**Comment by thefirebanks** [comment]
Finally, don't forget to add any new packages you used to `requirements.txt` please!

File: tasks/fine_tuning_sbert/src/loops.py (hunk 4 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/loops.py:129]: I included output_path here so it didn't have to be another global variable :) 
We could make one constant output_path for all experiments where the model would be saved but that would be problematic if more than one person saves a model on the google drive folder at the same time. I think it needs to be an input from the notebook for now.
It's not optimized over because we only pass in 1 value to the sweep config for output_path 😁

File: tasks/fine_tuning_sbert/src/loops.py (hunk 6 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/loops.py:228]: Yeah I kept thinking and found this to be a way to communicate the dataset between the notebook and the ```src``` files :D But for sure we can come up with a better solution for further experiments

File: tasks/fine_tuning_sbert/src/loops.py (hunk 7 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/loops.py:321]: I think that the CM and embedding visualization for the test set (inside the ``evaluate_using_sklearn`` function) can be stored in wandb. I made the adjustment for that and will test it. For the classification report (F1 accuracies), I think it is better to see and keep it in the notebook cell run itself.

File: tasks/fine_tuning_sbert/src/sentence_transformer.py (hunk 8 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/sentence_transformer.py:33]: Great idea! I made the adjustments to include this for sweeps and for a singular run 😃

File: tasks/fine_tuning_sbert/src/sentence_transformer.py (hunk 9 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/sentence_transformer.py:222]: I realize that our ``evaluator`` parameter is never None so this clause is obsolete. Because the ``save`` function for the model **only** happens in the ``_eval_during_training`` method.
I will remove this clause.

File: tasks/fine_tuning_sbert/src/sentence_transformer.py (hunk 10 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/src/sentence_transformer.py:244]: The validation accuracy values are stored in wandb one epoch at a time individually :) the ``ACC_LIST`` is only for monitoring the training in the notebook while it happens.
Do you think we should remove this now since the accuracy is stored in wandb? 🤔

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 11 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1788]: Ah yes! Sorry about leaving those there. Removed in the next push.

For spacy, I think we use it because the ``src`` python files need it for preprocessing the sentences. @jordiplanascuchi do you think we need spacy every time we run the notebook for fine tuning?

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 12 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1965]: We do delete the folder yes. But I'm concerned if 2 people running the notebook at the same time will get the same random_number, then both run's models will be saved in the same place. If it's a low probability, we can totally make it a random number.

I merged the cells too, they didn't need to be separate.

Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1965]: Also shifted the cell to remove the folder a little above so it's not missed while running everything

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 13 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2086]: Added the comments for this 😉

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 15 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2190]: Oh yeah that's true 😅 I will change the name for it

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 14 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:2104]: That's a good idea! Adding the comments now for clarity

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 11 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb:1788]: Actually, I just checked this and spacy needs to be there otherwise the repo import fails in the notebook

**Comment by ramanshsharma2806** [comment]
> 2 final things:
> 
> * Don't forget to add any new packages you used to `requirements.txt` please!
> * If we don't use spaCy by default, let's only initialize it in the function `evaluate_using_sbert` (as opposed to at the top of the `loops.py` file). That way we don't have to install it in the collab notebook every time :)

I will add the requirements in the txt file.
I think you're right with the spacy module being only used in the ``evaluate_using_sbert`` function. I will move the installation there so we don't have to install it in the notebook every time.

File: tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb (hunk 11 above)
Comment (ramanshsharma2806) [reply tasks/fine_tuning_sbert/notebooks/Fine_tuning_SBERT_with_callback.ipynb]: Actually it doesn't 😅 moved spacy installation to a sbert evaluation function

File: tasks/fine_tuning_sbert/src/loops.py (hunk 4 above)
Comment (thefirebanks) [reply tasks/fine_tuning_sbert/src/loops.py:127]: Gotcha, thanks!

**Review by thefirebanks — COMMENTED** [review]
Spectacular, thank you for addressing these comments!

**Review by thefirebanks — APPROVED** [review]
